rts/rts.c

```diff
@@ -1699,23 +1710,26 @@ int main(int argc, char **argv) {
         num_threads++;
     }
 
-    // We start on index 1 for pretty names, like Worker 1 is first, not 0, plus
-    // when doing CPU affinity, we want to bind worker threads to CPU 1, so we
-    // can reuse idx for that
-    for(int idx = 1; idx <= num_wthreads; idx++) {
-        pthread_create(&threads[num_threads+idx], NULL, main_loop, (void*)idx);
+    int wthread_start = num_threads;
+    for(int idx = 0; idx < num_wthreads; idx++) {
+        pthread_create(&threads[num_threads+idx], NULL, main_loop, (void*)idx+1);
         // Note how we pin thread index + 1, so worker thread 0 is on CPU 1
         // We use CPU 0 for misc threads, like IO / mon etc
         if (cpu_pin) {
             CPU_ZERO(&cpu_set);
-            CPU_SET(idx, &cpu_set);
+            CPU_SET(idx+1, &cpu_set);
             pthread_setaffinity_np(threads[num_threads+idx], sizeof(cpu_set), &cpu_set);
         }
     }
     num_threads = num_threads + num_wthreads;
 
-    for(int idx = 0; idx <= num_threads; ++idx) {
+    // Only joining the worker threads, we don't care about gracefully shutting
+    // down IO / Mon threads
+    for(int idx = wthread_start; idx < num_threads; ++idx) {
         pthread_join(threads[idx], NULL);
+        pthread_mutex_lock(&sleep_lock);
+        pthread_cond_broadcast(&work_to_do);
```

Comment:
I do not really know why this is necessary but without it, I saw intermittent hangs in our CI tests on Mac OS which I could also reproduce locally on my Mac, though not on Linux.

I'm quite confident this current code works, I have tested hundreds of runs on my local Mac without hangs.

Reply:
@nordlander if you can tell me why the pthread_cond_broadcast in main_loop is not enough, I'm all ears...

Reply:
Hmm. I wonder if this is a possible scenario: worker threads 1 and 2 are busy, all the others are sleeping. Thread 1 executes a continuation that calls exit, so the flag is set and all the sleeping threads are woken up -- except #2 which is already awake! So all threads except #2 now terminate as they should, but #2 is unaware of all this and eventually goes to sleep when it finds the readyQ empty.

So it's probably necessary to check the exit flag before going to sleep. Maybe that's where the exit handling code really should be, and not at the head of the loop?

Reply:
I can't remove it from the head of the loop because that would mean a worker thread will continue running as long as there is something on the readyQ - we have to be able to get out of that loop, but you are right about the sleep stuff I think.